### PR TITLE
Forecasting Menu [Work in Progress]

### LINKED ISSUE #346: [FR] Prediction Models Accuracy or F-Score metrics?
We cannot compare the prediction models for our different use cases unless we have some sort of scoring metric for each.

This will improve the prediction piece of the project tremendously imho.

- A

### LINKED ISSUE #1088: [FR] Neural prediction ignores end date (no backtesting)
I installed it using conda using the preset provided. I use the git upstream.
I tried backtesting with the neural networks (lstm, rnn, mlp) by adding --end-date (-e) 2021-10-10 -d 30 but all of them seem to ignore the flag (seeing from the code --end-date isn't being passed anywhere). This makes it impossible to backtest using the neural networks since they'll just spit out the latest data. It works fine for all other prediction techniques. Monte carlo doesn't seem to provide any end-date (backtest) flag anyway which it definetely should aswell (comparing the frequency of the mc results with the actual price)

### LINKED ISSUE #1778: [IMPROVE] Improve error message when openbb.stocks.pred. is used
**What's the feature that should be improved?**
pred_api does not get imported when tensorflow is not installed.
As a result, when openbb.stocks.pred.  is used, it's hard to know why the API is not working as expected.

**Describe how you would like the feature improved**
Ideally, we should know that openbb.stocks.pred was not fully imported due to tensorflow dependency

### LINKED ISSUE #2085: [Bug] All Prediction features: When a non-price value is the target column, the table output is incorrectly assumed as a price.
The tables that are printed from prediction menu features incorrectly assumes everything is a price. This causes rounding to occur where it shouldn't in the raw tables output by all features in this menu. 

The $ is also not located in the correct position. `$1.00`, not, `1.00 $`

Screen Shot 2022-07-13 at 10 53 20 AM

Screen Shot 2022-07-13 at 10 52 38 AM

Screen Shot 2022-07-13 at 10 58 52 AM

Screen Shot 2022-07-13 at 10 57 53 AM

`Target Column: Returns`
Screen Shot 2022-07-13 at 10 59 42 AM

### LINKED ISSUE #2128: [Bug] Installer v1.6 - Crypto/chart: Error: the command 'chart' doesn't exist on the /crypto/ menu.
Screen Shot 2022-07-21 at 7 23 51 PM

```
2022 Jul 21, 22:16 (🦋) /crypto/ $ chart
usage: crypto {headlines,candle,load,find,prt,resources,ta,dd,ov,disc,onchain,defi,tools,nft,pred,qa,cls,home,about,h,?,help,q,quit,..,exit,r,reset,support}
crypto: error: argument cmd: invalid choice: 'chart' (choose from 'headlines', 'candle', 'load', 'find', 'prt', 'resources', 'ta', 'dd', 'ov', 'disc', 'onchain', 'defi', 'tools', 'nft', 'pred', 'qa', 'cls', 'home', 'about', 'h', '?', 'help', 'q', 'quit', '..', 'exit', 'r', 'reset', 'support')

The command 'chart' doesn't exist on the /crypto/ menu.
```

The `candle` command does exist and functions, however it lacks the functionality of the candle command in `/stocks`

```
2022 Jul 21, 22:22 (🦋) /crypto/ $ candle -h
usage: candle [-h] [--export EXPORT]

Display chart for loaded coin. You can specify currency vs which you want to show chart and also number of days to get data for.

optional arguments:
  -h, --help       show this help message (default: False)
  --export EXPORT  Export raw data into csv, json, xlsx and figure into png, jpg, pdf, svg (default: )

For more information and examples, use 'about candle' to access the related guide.
```

### PR DESCRIPTION
**Pull request by martinb-bb**
# Description

- [x] Forecasting menu option
- [x] Forecasting Controller/model/view
- [x] Show dataset
- [x] Plot dataset
- [x] Combine dataset
- [x] Delete dataset column
- [x] Rename dataset column
- [x] Load dataset
- [x] Probabilistic Exponential Smoothing model
- [x] Theta model
- [x] Probabilistic Recurrent Neural Network (RNN, LSTM, GRU)
- [x] Block Recurrent Neural Network (RNN, LSTM, GRU) (feat. Past covariates)
- [x] Neural Bayesian Estimation (feat. Past covariates)
- [x] Temporal Convolutional Neural Network (feat. Past covariates)
- [x] Dataset cleaning
- [x] Dataset descriptive statistics
- [x] Feature engineering menu option
- [x] EMA
- [x] Stochastic Oscillator
- [x] RSI
- [x] ROC
- [x] Momentum
- [x] % Delta change
- [x] Average true range
- [x] simple moving average short/long window
- [x] price signal
- [x] Refresh loaded dataset on menu
- [x] Regression model (@colin99d )
- [x] Transformer model (@colin99d - not temporal)
- [x] Temporal Fusion transformer (performs weirdly)
- [x] Display user compute device (CPU vs. GPU)
- [x] Display user torch and darts version
- [x] Port over Arima (non torch) (@colin99d ) (Hide for now)
- [x] Port over knn (non torch) (@colin99d )
- [x]  Port over MonteCarlo (non torch) (@colin99d)
- [x] Linear Regression (@colin99d )
- [x] Plot seasonality of data (@colin99d this technically "works" but graph looks terrible, reached out to @piiq )
    - Going to use our own version of their graphs for now, in the future might work with them to add functionality.
    - Also, submitted a PR to darts: https://github.com/unit8co/darts/pull/1056
- [x] Plot residuals from best_model (flag to set before training) (plot_residuals_analysis()) (@colin99d) --> Theta model does not work - make theta just print out that its not available. Also, theme needs a few tweaks. (@colin99d ) 
- [x] Plot dataset feature correlation coefficients
- [x] Pytest - Just make it run and successfully train (@colin99d )
- [x] Base class/ refactoring (@colin99d + @martinb-bb ) - 69.420% done for now...
- [x] Get rid of this printout:
`[2022-06-16 14:43:03,449] WARNING | darts.models | Support for AutoARIMA, BATS and TBATS is not available.To enable it, install u8darts[pmdarima] or u8darts[all].
[2022-06-16 14:43:03,450] WARNING | darts.models | Support for Facebook Prophet is not available. To enable it, install "darts", "u8darts[prophet]" or "u8darts[all]" (with pip);or "u8darts-all" (with conda).` (fixed with PR https://github.com/unit8co/darts/pull/1010) 
- [x] Remove external_axis from the docstrings or use the argument(@colin99d)
- [x] Glossary (@colin99d) (dict)
- [x] Confirm features work with Polygon and cryptodata (@colin99d)
    - For this PR we support daily, weekly, and monthly
    - https://github.com/unit8co/darts/issues/1057
- [x] Check if ram requirements match data size
- [x] plot only historical forecast (not entire series) (@colin99d)
- [x] Develop API (@colin99d)
- [x] Remove pred menus from the terminal (@colin99d )
- [x] Remove Tensorflow requirements  (@colin99d )
- [x] Allow to bring any data from another menu to forecasting directly  `stocks/load AAPL/forecast/USE MODEL` - works for both stocks and crypto menu. - to add economy soon or in V2
- [x] grab exported data from EXPORT folder as well (@colin99d)
- [x] standarize all models/views for ease of use in API
- [x] load mutliple stocks from stocks menu
- [x] Give combine warnings if it needs to use index
- [x] Add --all-covariates flag
- [x]  Naive Bayes (for setting baseline)
- [x] If someone doesn't specify combine columns combine all columns
- [x] Rename menu to forecast
- [x] Make the disclaimer when we boot up terminal
- [x] Handle no darts gracefully
- [x] Fix hard-coded export path (this looks like a Darren specific issue)
- [x] Do not format non dollar amounts as dollar amounts
- [x] Standardize names in columns (default string for ticker)
- [x] Have a better way of handling target column not working

Other Bugs Fixed:
- Crypto export command does nothing
- Fixes #2128
- Fixes #2085
- Fixes #1088
- Fixes #1778 
- Fixes #346 

# Checklist:

- [x] Update [our Hugo documentation](https://openbb-finance.github.io/OpenBBTerminal/) following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/website).
- [x] Update our tests following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/tests).
- [x] Make sure you are following our [CONTRIBUTING guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/blob/main/CONTRIBUTING.md).
- [x] If a feature was added make sure to add it to the corresponding [scripts file](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/scripts).
- [x] One letter parameters for common args
- [x] Handle regular days included with Business days when building timeseries
- [x] make disclaimer for loading menu
- [x] remove default csvs (@colin99d)
- [x] Notebooks with usage examples

Known issues
- [x] rnn residuals with aapl.csv does not work - fixed: https://github.com/unit8co/darts/pull/1066 to be added in 0.21.0
- [x] reset command (@colin99d )
- Make reset work when loading from another menu (im making this out of scope for now, if anyone legitimately wants this feature we can add in V2)
- [x] positive series with MAPE bug (https://github.com/unit8co/darts/issues/1063) - fixed
- [x] linregr and regr prints a lot of `FutureWarning: pandas.Int64Index is deprecated` warnings
- [x] Plot shows numbers instead of dates for X axis
- [x] Big dataframes print in an ugly fashion
Screen Shot 2022-07-21 at 7 22 40 PM

V2 Features:
- Bring in Darts Datasets for further user experimentation
- Plot data in different ways (bar chart, box and whisker, 3D???)
- multi input and 2 y axis plotting ( economy menu) 
- Static covariates
- stochastic forecasts capturing model uncertainty using Monte Carlo Dropout with deep learning models
- Load previously trained models for inference
- Train OpenBB NN model for off-the-shelf forecasting ( Long term goal)) 
- past covariates with different delta T
- Create a dashboard
- encoding string variables in columns
- hyperparameter tuning with Raytune: https://docs.ray.io/en/latest/tune/examples/tune-pytorch-lightning.html
- Do a better job at merging dataframes with different times (Polygon data can be 4 hours off)
- Refactor the views to use all the same code
- More descriptive feature engineering names (running the same feat eng on different columns replaces old values)

# Others
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My code passes all the checks pylint, flake8, black, ... To speed up development you should run `pre-commit install`.
- [x] New and existing unit tests pass locally with my changes. You can test this locally using `pytest tests/...`.

To use in Jupyter:
from openbb_terminal.api import forecast as fc
from openbb_terminal import api

import matplotlib.pyplot as plt
import matplotlib_inline.backend_inline

%matplotlib inline
matplotlib_inline.backend_inline.set_matplotlib_formats("svg")
api.theme.applyMPLstyle()

### CODE REVIEW

**Comment by DidierRLopes** [comment]
Someone has been avoiding sleep 👀

**Comment by jmaslek** [comment]
This is looking awesome.  Small point I'm noticing.  For controllers/argparse choices, y'all are doing `--some_thing`.  

We should stick to doing `--some-thing`

**Comment by martinb-bb** [comment]
Gotcha, will update!

**Comment by deeleeramone** [comment]
https://github.com/OpenBB-finance/OpenBBTerminal/pull/2208 - FYI

File: openbb_terminal/dashboards/dashboards_controller.py [hunk 1]
```diff
@@ -129,13 +138,25 @@ def create_call(other_args: List[str], name: str, filename: str = None) -> None:
         dest="dark",
         help="Whether to show voila in dark mode",
     )
-
+    if streamlit:
```
Comment (piiq) [inline openbb_terminal/dashboards/dashboards_controller.py:141]: In this controller we use 
```python
 subprocess.Popen(
                f"{cmd} {file} {args}",
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                shell=True,
            )
```
which raises a security warning in addition to the fact that this can't be launched from the installer version.

make sure you know how to launch streamlit from the terminal python loop without using the `subprocess` module.
Someone on the team is already working on making the Voila launch from inside the terminal in order to make it available to the users of the installer version

Comment (piiq) [reply openbb_terminal/dashboards/dashboards_controller.py:141]: if you do `which streamlit` and then `cat` or open that file, you'll see that it's a regular python script that launches some main loop. We can launch the same main loop but from inside the terminal

this is `streamlit` launcher
```
# -*- coding: utf-8 -*-
import re
import sys
from streamlit.web.cli import main
if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(main())
```

**Comment by JerBouma** [comment]
@martinb-bb Could you write an "Introduction to" page for the Forecasting menu? See as an example the following: https://openbb-finance.github.io/OpenBBTerminal/terminal/stocks/ or https://openbb-finance.github.io/OpenBBTerminal/terminal/economy/.

This is crucial so users can get the hang of how the menu works. We have functionality within the terminal (`about`) that links the user automatically to the page:

image

**Comment by martinb-bb** [comment]
Yes, working on it!

File: openbb_terminal/forecast/forecast_view.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    theme.style_primary_axis(ax)
+
+    if external_axes is None:
+        theme.visualize_output()
+
+    export_data(export, os.path.dirname(os.path.abspath(__file__)), "plot")
+
+
+@log_start_end(log=logger)
+def show_df(df: pd.DataFrame, limit: int = 15, name: str = "", export: str = ""):
```
Comment (minhhoang1023) [inline openbb_terminal/forecast/forecast_view.py:215]: To conform to the Contributing guide, this should be `data` instead of `df`

File: openbb_terminal/forecast/forecast_view.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        headers=list(df.columns),
+        show_index=True,
+        title=f"Dataset {name} | Showing {limit} of {len(df)} rows",
+    )
+
+    export_data(export, os.path.dirname(os.path.abspath(__file__)), f"{name}_show", df)
+
+
+@log_start_end(log=logger)
+def describe_df(df: pd.DataFrame, name: str = "", export: str = ""):
```
Comment (minhhoang1023) [inline openbb_terminal/forecast/forecast_view.py:234]: Similar to the above, this should be data instead of df

**Review by minhhoang1023 — DISMISSED** [review]
Left some minor comments about the argument naming convention.

File: openbb_terminal/dashboards/stream/forecast.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+            # TODO need to new add in business days and merge to make new DF before plotting
+            # create df with 5 next business days from a start date
+            # next_business_days = pd.date_range(start=end_date, periods=5, freq="B")
+
+        # if st.button("Add Column"):
+        # kwargs = {}
+        # if has_parameter(handler.feature_model, "target_column"):
+        # kwargs["target_column"] = handler.feature_target
+        # handler.df = handler.feature_model(handler.df, **kwargs)
+        # st.session_state["widget_options"]["past_covs_widget"] = handler.df.columns
```
Comment (piiq) [inline openbb_terminal/dashboards/stream/forecast.py:298]: @colin99d should be here?

**Review by piiq — CHANGES_REQUESTED** [review]
We have predict menu leftovers here and there, is it supposed to be like that?

image

image

image

image

File: openbb_terminal/dashboards/stream/forecast.py (hunk 4 above)
Comment (colin99d) [reply openbb_terminal/dashboards/stream/forecast.py:298]: Yeah. I will turn that into working code later.

**Review by piiq — APPROVED** [review]
LFGO 🚀

**Review by jmaslek — APPROVED** [review]
I want to join in on the fun
